File: Java/search/TernarySearch.java
import java.util.*;
public class Main
{  
   // Recursive Approach
  static int ternarySearchRecursive (int arr[], int f, int left, int right)
  {
    if (left <= right)
      {
	int mid1 = left + (right - left) / 3;
	int mid2 = right - (right - left) / 3;
	if (arr[mid1] == f)
	  return mid1 + 1;
	else if (arr[mid2] == f)
	  return mid2 + 1;
	else if (f < arr[mid1])
	    return ternarySearchRecursive (arr, f, left, mid1 - 1);
	else if (f > arr[mid2])
	    return ternarySearchRecursive (arr, f, mid2 + 1, right);
	else

	    return ternarySearchRecursive (arr, f, mid1 + 1, mid2 - 1);
      }
    return -1;
  }
    // Iterative Approach
 static int ternarySearch(int arr[], int f, int left, int right)
    {  int found = -1;
        while(left <= right)
        {
          int mid1 = left + (right - left) / 3;
	int mid2 = right - (right - left) / 3;
            if(arr[mid1] == f)
                return mid1 + 1;
            else if(arr[mid2] == f)
                return mid2 + 1;
            else if(f < arr[mid1])
             right = mid1 - 1;
            else if(f > arr[mid2])
               left = mid2 + 1;
            else 
              {
                  left = mid2 + 1;
                  right = mid2 - 1;
              }
            
        } 
        return found;
       
    }
    

  public static void main (String[]args)
  {
    Scanner sc = new Scanner (System.in);
    int i, n, left, right, f = -1, element;
    System.out.println ("Enter length of array");
    n = sc.nextInt ();
    System.out.println ("Enter  array elements");
    int arr[] = new int[n];
    for (i = 0; i < n; i++)
      arr[i] = sc.nextInt ();
    System.out.println ("Enter element to be searched");
    element = sc.nextInt ();

    Arrays.sort (arr);
    left = 0;
    right = n - 1;

    f = ternarySearch(arr, element, left, right);
    if (f != -1)
      System.out.println ("Element found at " + (f));
    else
      System.out.println ("Element not found");
  }
}

/*
Time Complexity -  O(log3(N))

Sample I/O :

Input :  
Enter length of array
4
Enter  array elements
1 2 3 4
Enter element to be searched
4

Output :
Element found at 4 
*/

File: Kotlin/search/TernarySearch.kt
//Kotlin program to implement ternary search using recursive approach

import java.util.*

// A function to declare ternary search
fun ternarySearch(left:Int, right:Int, key:Int, array: ArrayList<Int>): Int{
    if (right >= left) {

        // Finding the midterms
        val mid1 = left + (right - left) / 3
        val mid2 = right - (right - left) / 3

        // Check if the value is present in the first mid term
        if (array[mid1] == key) {
            return mid1
        }

        // Check if the value is present in the second mid term
        if (array[mid2] == key) {
            return mid2
        }

        // If the element is not present in the mid positions, the following cases can be a possibility.

        // Checking if the value is less than mid1 element
        if (key < array[mid1]) {

            return ternarySearch(left, mid1 - 1, key, array)
        }

        // Checking if the value is greater than mid2 element
        else if (key > array[mid2]) {
            return ternarySearch(mid2 + 1, right, key, array)
        }

        // The last possibility is that, the element may be present between the mid1 and mid2
        else {
            return ternarySearch(mid1 + 1, mid2 - 1, key, array)
        }
    }
    // If all the possibilities fail, the element is not present inside the array
    return -1
    }

//Testing code
fun main(){

    val input = Scanner(System.`in`)

    println("Enter the length of the array")
    val arrayLength = input.nextInt()
    val arrayOfElements = arrayListOf<Int>()

    println("Enter the elements of the array in ascending order")
    for(index in 0 until arrayLength) {
        val element = input.nextInt()
        arrayOfElements.add(element)
    }

    print("Enter the number you want to search for :")
    val number = input.nextInt()
    input.close()
    val left:Int = -1
    val right:Int = arrayLength - 1

    // Search the number using ternarySearch
    val position = ternarySearch(left, right, number, arrayOfElements) + 1

    if(position == 0)
        println("Key: $number is not present in the array")
    else
        println("Key: $number is present in the array at position $position")
}

/*
    Sample Test Cases:-

    Test Case 1:-
    Enter the length of the array
    5
    Enter the elements of the array in ascending order
    1 2 3 4 5
    Enter the number you want to search for :5
    Key: 5 is present in the array at position 5

    Test Case 2:-
    Enter the length of the array
    5
    Enter the elements of the array in ascending order
    10 20 30 40 50
    Enter the number you want to search for :60
    Key: 60 is not present in the array

    Time Complexity:- The time complexity of this algorithm is O(log3n), where n is the size of the array
    Space cComplexity:- The space complexity of this algorithm is O(1), which is constant, irrespective of any case.
 */
